C++: Add support for `Element` content

MathiasVP opened the pull request:
This PR adds support for `Element` content for C++ like we have for other CodeQL supported languages. Given code such as:
```cpp
std::vector<int> v;
int x = source();
v.push_back(x);
// ...
int y = v.at(i);
sink(y);
```
we now track the flow into `v` by a store step that pushes `Element` onto the access path and transfers flow to `v`. Then, once we reach `v.at(i)` we pop `Element` off the access path and transfer flow to the return value. This means we now track container flow as _data_ flow instead of taint flow.

This is all very standard stuff for other languages by now (Tom did this for C# [almost four years ago](https://github.com/github/codeql/pull/3366)), but there are (of course 😂) a few challenges for C++. Some of these I've resolved in this PR, and I'll create follow-up issues for the rest.

Commit-by-commit review recommended.

## Signatures for templates

How do we distinguish between these two constructors in a MaD summary (from [here](https://en.cppreference.com/w/cpp/container/vector/vector)):
- `explicit vector(size_type count);`
- `explicit vector(size_type count, const T& value = T(), const Allocator& alloc);`

the obvious answer in MaD is to distinguish them by providing a signature. However, the signature for the latter constructor mentions the parameter `T` whose parameter name is given by a template (i.e., the enclosing class' template list). And we can't be sure that this template is named `T` across all implementations of the STL.

Likewise, consider this constructor:
```cpp
template< class InputIt > vector( InputIt first, InputIt last, const Allocator& alloc);
```
whose parameters are given by the template list of the function itself.

To solve the above problem, I've introduced MaD syntax for introducing template parameters when specifying the class name and the function name. Staying true to the spirit of C++, I've chosen this syntax:
```yml
["std", "vector<T,Allocator>", _, "vector<InputIterator>", "(InputIterator,InputIterator,const Allocator &)", "", _, _, _, _]
```
As you can see, the signature mentions both `Allocator` (from the class template), and `InputIterator` (from the function template).

This ensures that this MaD row will match the right constructor no matter what the implementation picked as template parameter names.

Internally, the name `(InputIterator,InputIterator,const Allocator &)` is normalized to a name that doesn't mention the template parameters before being compared with a normalized version of the parameter list in the database.

## Handling indirections

The obvious summary for a function like [push_back](https://en.cppreference.com/w/cpp/container/vector/push_back) is: `Argument[*0] -> Argument[-1].Element[*]`.

That is, it reads the value pointed to by the 0'th argument (because `push_back` takes a parameter by reference), and stores it as an `Element` content on the qualifier.

However, consider this example:
```cpp
std::vector<int*> v;
int x = source();
int* p = &x;
v.push_back(p);
...
int* q = v.at(i);
sink(*q);
```
In this case, we track `*p` (i.e., the indirection of `p`) when we perform `push_back`, which means that the value that actually flows into `push_back` is the address of `*p`. That is, the node reresenting `**p` is actually what ends up flowing to the 0'th argument of `push_back`.

To solve this, we could introduce another summary for `push_back`: `Argument[**0] -> Argument[-1].Element[**]`.

However, that introduces quite a burden on the developer when adding MaD rows. For example, if we need to add an additional indirection to every model we need to introduce another MaD row to _every function_.

To solve this problem I introduced a kind of "template" (no pun intended) for specifying an arbitrary number of indirections. So the summary for `push_back` actually looks like: `Argument[*@0] -> Argument[-1].Element[*@]` where `@` means (any number of indirections).

When consuming MaD rows we then expand the summaries to:
- `Argument[*0] -> Argument[-1].Element[*]`
- `Argument[**0] -> Argument[-1].Element[**]`
- `Argument[***0] -> Argument[-1].Element[***]`
- `Argument[****0] -> Argument[-1].Element[****]`
- `Argument[*****0] -> Argument[-1].Element[*****]`

And I've capped the number to `5`. Dataflow performance wise there should be no concerns regarding pushing this further up, but it does mean we produce a lot more strings. We could also consider capping this to
"the maximum number of indirections in the database". I'll let this be future work, though :joy:

I'm not super happy with the syntax I've chosen here, so I'm happy to bikeshed on this once we're happy with everything else in this PR.

## Supporting associative containers

Initially, I wanted to also replace all of our taint models for `std::map` with MaD rows. However, I ran into a performance problem. To see what happens, consider the hypothetical MaD summary for [std::map::insert](https://en.cppreference.com/w/cpp/container/map/insert):
```cpp
std::pair<iterator, bool> insert( const value_type& value );
```
(ignoring `@`s for this example since it's not relevant for this discussion): `Argument[*0] -> ReturnValue.Field[first].Element[*]`.

That is, the input is the pointer to the first argument (since insert takes a const reference), and returns a std::pair whose first element is an iterator to the inserted element.

Now: Which exact field does `Field[first]` refer to? Since `std::pair<T, U>` is a template there will be many many `std::pair<T, U>` instantiations in the database, and the above MaD row will create a store step that writes any of them to the access path. On a DB I was testing with there were ~1500 instantiations of `std::pair<T, U>` (i.e., ~1500 combinations of `T` and `U`) which made the flow summary library really unhappy.

To solve this problem we need to extend the MaD language so that we can annotate which `first` field we're talking about (i.e., it's the `first` field of the `std::pair` class whose template parameters are `T = the template argument from the enclosing std::map` and `U = bool`). I haven't done this yet, though.

To "solve" this performance problem I've simply chosen not to model any associative containers for now. However, anyone who ventures into doing this will hit the same performance problem. So we need to resolve this at some point (but not necessarily in this PR since this is large enough as-is).

MathiasVP commented:
DCA shows nothing spectacular:
- Performance is fine (this was my main worry 🎉)
- One lost result which has to do with iterators still not being fully converted to use `Element` content. I expect this result to appear once we model iterators fully with `Element` content.
- Two new FPs on `cpp/double-free`. These appear because `Element` content allows us to have dataflow through containers now (instead of simply having it be taintflow), so we need to restrict container flow in `cpp/double-free` similarly to how we [restrict flow through ArrayExpr](https://github.com/github/codeql/blob/6dbdc9e17fb5a6640c14a0b586570c1c37fe7920/cpp/ql/lib/semmle/code/cpp/security/flowafterfree/FlowAfterFree.qll#L98) in that query.

geoffw0 commented:
I don't really understand the motivation for `@`.  If we're not propagating indirect data properly through models, isn't that an issue with data-flow itself (or the indirections feature) - not an issue with the MAD representation?

MathiasVP commented:
> I don't really understand the motivation for `@`. If we're not propagating indirect data properly through models, isn't that an issue with data-flow itself (or the indirections feature) - not an issue with the MAD representation?

In general we don't infer a `*n1 -> *n2` step when a model specifies a `n1 -> n2` step. We actually used to do this (see [this PR](https://github.com/github/codeql/pull/5035)), but pulled it out pretty quickly because of _your_ excellent [comment](https://github.com/github/codeql/pull/5035#issuecomment-769171380) on the PR.

geoffw0 commented:
Thanks, that makes sense.  I'll find time to review the code changes at a later point, please ping me if anything is waiting on this.

MathiasVP commented:
Sounds good. Thanks! I'm not blocked on getting this merged so feel fee to take all the time you need 😄

geoffw0 commented on cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll:
```diff
@@ -396,6 +567,150 @@ private predicate matchesSignature(Function func, string signature) {
+  beforeAngles = s.regexpCapture("([^<]+)(?:<([^>]+)>(.*))?", 1) and
+  (
+    betweenAngles = s.regexpCapture("([^<]+)(?:<([^>]+)>(.*))?", 2) and
+    afterAngles = s.regexpCapture("([^<]+)(?:<([^>]+)>(.*))?", 3)
```
Does it matter if we use `[^>]+` or `[^>]*`?  (I have no preference if both are correct)

MathiasVP replied:
By having `[^>]+` we disallow writing `<>`. That is, if we have `[^>]*` we'll allow a function name such as `foo<>`, whereas now we require that to be `foo`. I'd like to have _one way_ of writing this out, which is why I decided to go with `[^>]+`

geoffw0 commented on cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll:
```diff
@@ -396,6 +567,150 @@ private predicate matchesSignature(Function func, string signature) {
+    or
+    not exists(s.regexpCapture("([^<]+)(?:<([^>]+)>(.*))?", 2)) and
+    betweenAngles = "" and
+    afterAngles = ""
```
What case does this cover?

MathiasVP replied:
This case from the QLDoc:
```
or `s = beforeAngles` where `beforeAngles` does not have any brackets.
```
That is, there's no angle brackets in the string so we cannot capture anything in between the angle brackets. In that case, `betweenAngles` and `afterAngles` are empty, and `beforeAngles` is equal to `s`.

geoffw0 replied:
I see, it's not intended to cover other cases like `foo<>` (discussed above), just the case where there are no angle brackets whatsoever (or to be more precise: no matching angle brackets whatsoever).

geoffw0 reviewed (commented):
Initial review.  There's a lot going on here, so lots of comments, sorry.

I've reviewed most of the QL changes, skimmed some parts though, and looked at the DCA run as well.  I have not yet reviewed the new models or the impact on tests.

I don't agree with the `no-change-note-required` tag.  I think we need to explain the changes to the MAD syntax to users.

MathiasVP commented:
> I don't agree with the `no-change-note-required` tag. I think we need to explain the changes to the MAD syntax to users.

That's fair. I wanted to delay the change note until we were sure this was the format we wanted. Note that the syntax is backwards compatible with everything we had before. Once we're happy with this syntax I'll amend our existing documentation and add a proper change note (which may happen in a subsequent PR if that's okay with you).

MathiasVP commented on cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/TaintTrackingUtil.qll:
```diff
@@ -147,7 +147,10 @@ predicate defaultAdditionalTaintStep(DataFlow::Node src, DataFlow::Node sink, st
+predicate defaultImplicitTaintRead(DataFlow::Node node, DataFlow::ContentSet c) {
+  node instanceof ArgumentNode and
+  c.isSingleton(any(ElementContent ec))
+}
```
The reason we want this is for the same reason as we have it in Swift: https://github.com/github/codeql/blob/9d8ee99c1caafb2ee02e760f76b9e894e619094d/swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPublic.qll#L45-L48

I went a bit on the safer side and put in a restriction similar to what Java has: https://github.com/github/codeql/blob/9d8ee99c1caafb2ee02e760f76b9e894e619094d/java/ql/lib/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll#L183. That is, they only have the implicit read step at arguments. We could remove this restriction if you want.

MathiasVP commented on cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll:
```diff
@@ -367,16 +400,154 @@ private predicate elementSpec(
 
-private string paramsStringPart(Function c, int i) {
-  i = -1 and result = "(" and exists(c)
+/** Gets the fully templated version of `f`. */
```
Yes, exactly. This is harder than it sounds because we also need the function belonging to a _class_ template. For example:
```cpp
template<typename T>
class C {
  void f(const T& t);
};
``` 
the `f` is not a function template, but it is a member function of the enclosing class template. And we need to find _that_ function.

MathiasVP commented on cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll:
```diff
@@ -396,6 +567,150 @@ private predicate matchesSignature(Function func, string signature) {
+private predicate parseAngles(
+  string s, string beforeAngles, string betweenAngles, string afterAngles
+) {
+  beforeAngles = s.regexpCapture("([^<]+)(?:<([^>]+)>(.*))?", 1) and
```
Ah, that's part of the syntax that I didn't spell out. The following isn't a valid signature:
```
(const vector<T> &)
```
for exactly the reason you've written it: It wouldn't be easy to parse. So the signature you specify in the MaD row cannot include templates. Instead, you'll have to write it as:
```
(const vector &)
```
I'll add a comment at the appropriate place for this.

MathiasVP replied:
For example, you can see a use of such a signature in the copy constructor for vector here: https://github.com/github/codeql/pull/16791/files#diff-4ac8e1a72ed803d7de0e342cfe1bbc7cd0cd05c6618ed33a837c5c0d6a1ea78eR45 (note the missing angle brackets around `vector`).

MathiasVP replied:
> I'll add a comment at the appropriate place for this.

Done in d38ce61e032c7b393b434f43fa746465ed9f85e2

geoffw0 replied:
OK.  So what about `parseAngles`, should there only ever be one set of `<>` in the input string `s`?

MathiasVP replied:
`parseAngles` is used to destruct the value in the `type` and `name` columns. If you have multiple `<>` in a string that means that either `type` looks like: `myType<abc<foo>>` (or similar for `name`), and that's not a valid value for the column since it's supposed to be of the form `myType<A,B,C>` (i.e., a name followed by template parameter names).

MathiasVP replied:
So TLDR: Yes.

MathiasVP commented on cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll:
```diff
@@ -367,16 +367,154 @@ private predicate elementSpec(
+    tp = templateFunction.getTemplateArgument(remaining) and
+    result = mid.replaceAll(tp.getName(), "func:" + remaining.toString())
+  )
+}
```
Sure. They're being tested through our existing taint flow tests, but I can add unit tests for it if that will make you more comfortable.

MathiasVP replied:
e845204ec962fa18595a1dfd2fed95bceb6019e5 adds test for this. I hope that's the kind of thing you were thinking of? If not, please let me know and I'll be happy to do something else!

geoffw0 replied:
Thanks.  The new tests certainly shine light on the results of these predicates, and will make any future changes to them clear.  I still don't understand the _intent_ behind these three predicates though.

MathiasVP replied:
I've corrected and expanded the QLDoc in c104a0a74c6d272b1423ffec3276031f73ec02b4.

The intent behind `signatureMatches` is that we want to compare the signature _as written in a MaD row_ with the signature of a function _in the database_.

In order to do that, we need to:
- Find the i'th entry in the `signature` column in the MaD row (that's the job of `getSignatureParameterName`)
- Find the i'th parameter of the function in the database and compute its string representation (that's the job of `getParameterTypeName`)
- Compare those two values (that's the job of `signatureMatches`)

Obviously, every type in the signature must match the function, so naturally `signatureMatches` is recursive.

geoffw0 replied:
Thanks, the example in the new QLDoc really helped. :tada:

MathiasVP commented:
> Initial review. There's a lot going on here, so lots of comments, sorry.

No need to apologise for having many comments 😄

MathiasVP commented on cpp/ql/test/library-tests/dataflow/dataflow-tests/type-bugs.expected:
```diff
@@ -1,5 +1,37 @@
 astTypeBugs
 irTypeBugs
+| ../../../include/iterator.h:21:3:21:10 | ../../../include/iterator.h:21:3:21:10 | ../../../include/iterator.h:21:3:21:10 | [summary param] *0 in iterator |
```
I'm not totally sure what's going on with these tests. It may be related to how we model iterators in dataflow (i.e., we have special code to recognize `*it = x` as being a write to the container that created the iterator `it`). I don't think we need to fix this in this PR.

MathiasVP replied:
Indeed. In fact, either the problem is that it has more than one associated type, or that it has no associated type: https://github.com/github/codeql/blob/ce177c345033595a41e9eb9931fda65d15cf887e/cpp/ql/test/library-tests/dataflow/dataflow-tests/type-bugs.ql#L21

We should add a `description` to this query predicate to make it slightly more clear what the problem is.

In any case, I plan on investigating this as a follow-up if you're okay with it. It shouldn't have any affect on query results.

geoffw0 replied:
Sounds good to me.

geoffw0 commented on cpp/ql/lib/semmle/code/cpp/models/implementations/StdContainer.qll:
```diff
@@ -55,73 +55,6 @@ private class Vector extends StdSequenceContainer {
-    output.isQualifierObject()
-  }
-
-  override predicate isPartialWrite(FunctionOutput output) { output.isQualifierObject() }
```
Is the information from `isPartialWrite` in the old models encoded in the new models somewhere?

MathiasVP replied:
That happens automatically due to the way field flow works (the same is true for all CodeQL languages). Consider an example such as:
```cpp
1. std::vector<int> v;
2. v.push_back(source());
3. v.push_back(0);
4. sink(v.at(0));
```
on `main` the [modeledFlowBarrier](https://github.com/github/codeql/blob/40b75342104ec6a449750332af0d81a501cf9ce0/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll#L802) predicate excludes cases where `isPartialWrite` holds. However, now that this is tracked via field flow, what happens is that:

```mermaid
graph TD;
    A["source()"]-->|storeStep| B["v [post update] on line 2"];
    B-->|SSA| C["v on line 3"];
    C-->|SSA| D["v on line 4"];
    E["0"] -->|storeStep| G["v [post update] on line 3"];
    G -->|SSA| D;
    D -->|readStep| F["v.at(0)"];
```

and this happens completely automatically via the existing interaction between SSA and field flow.

geoffw0 commented on cpp/ql/lib/ext/std.array.model.yml:
```diff
@@ -0,0 +1,12 @@
+      pack: codeql/cpp-all
+      extensible: summaryModel
+    data: # namespace, type, subtypes, name, signature, ext, input, output, kind, provenance
+      - ["std", "array", True, "at", "", "", "Argument[-1].Element[*@]", "ReturnValue[*@]", "value", "manual"]
```
Why are all the models using `Element[*]` (or `Element[*@]`) and not just `Element[]` (or `Element[@]`)?  Surely `array.at(0)` contains elements, not element references?

MathiasVP replied:
Both `vector::at` and `vector::operator[]` return references to their elements: https://en.cppreference.com/w/cpp/container/vector/at

This is why you can do stuff like:
```cpp
std::vector<int> v{10};
v.at(0) = 42;
```

geoffw0 replied:
Yep, the return value is a reference, but the input `Argument[-1].Element[*@]` surprised me.  Am I reading this right as "one or more dereferences of an element contained in the qualifier"?  What are the dereferences about?

MathiasVP replied:
Because you can also do:
```cpp
std::vector<int*> v{10};
// ...
*v.at(0) = x;
```
That is, it's not the content that's tainted, it's the value of the pointed to content.

MathiasVP replied:
Or equivalently, you can do:
```cpp
std::vector<int*> v{10};
int x = source();
v.at(0) = &x;
```

MathiasVP replied:
You are absolutely right. In https://github.com/github/codeql/pull/16791/commits/5be948533c22525716758721635fe69095c11492 I've replaced all mentions of `Element[*@]` with `Element[@]`.

geoffw0 commented on cpp/ql/lib/ext/std.array.model.yml:
```diff
@@ -0,0 +1,12 @@
+      - ["std", "array", True, "data", "", "", "Argument[-1].Element[*@]", "ReturnValue[*@]", "value", "manual"]
+      - ["std", "array", True, "operator[]", "", "", "Argument[-1].Element[*@]", "ReturnValue[*@]", "value", "manual"]
+      - ["std", "array", True, "rbegin", "", "", "Argument[-1].Element[*@]", "ReturnValue.Element[*@]", "value", "manual"]
+      - ["std", "array", True, "rcbegin", "", "", "Argument[-1].Element[*@]", "ReturnValue.Element[*@]", "value", "manual"]
```
This is probably the last thing you should fix (after other changes), but all of the `std` models included `bsl` equivalents as well.

MathiasVP replied:
Good point. I've added these in 2ad8704325fe740f0ae43b987ae9d86993732fe9. I didn't find any `bsl` version of iterators, so I've left those out for now.

MathiasVP replied:
Did another DCA run after https://github.com/github/codeql/commit/2ad8704325fe740f0ae43b987ae9d86993732fe9, and results are unchanged

geoffw0 replied:
:+1:

The `StdIterator`, `IteratorTraits` and `IteratorByTypedefs` QL classes do cover `bsl`.  None of these classes were removed by your changes, however.

geoffw0 commented:
> I wanted to delay the change note until we were sure this was the format we wanted.

That sounds sensible.  There's also the information at the top of `ExternalFlow.qll` that needs to be updated with the new syntax (`Element` and `@`).

MathiasVP commented:
> That sounds sensible. There's also the information at the top of `ExternalFlow.qll` that needs to be updated with the new syntax (`Element` and `@`).

I've done (a version of) this in 0b4459db7587c20acf3219dc50f43f4a8968f42a:
- I've added QLDoc to mention how to introduce template names from class names and function names,
- I've added QLDoc to mention `@`

I have _not_ added QLDoc for `Element` content since we don't have QLDoc for `Field` content, nor do any other language have this. It should probably be done eventually, but I'd prefer to delay this.

geoffw0 commented on cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -1700,6 +1709,14 @@ class DataFlowSecondLevelScope extends TDataFlowSecondLevelScope {
+ * This should be equal to the largest number of stars (i.e., `*`s) in any
+ * `Element` content across all of our MaD summaries, sources, and sinks.
+ */
+int getMaxElementContentIndirectionIndex() { result = 5 }
```
Might be best to do this as follow-up, with an eye on variations to performance?

MathiasVP replied:
Yeah, I agree 🙂

geoffw0 commented on cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll:
```diff
@@ -65,6 +74,17 @@
+ *    - flow from the _first_ indirection of the 0th argument to the return
+ *    value, and
+ *    - flow from the _second_ indirection of the 0th argument to the first
+ *    indirection of the return value, etc.
```
This is a helpful detail :+1:

geoffw0 commented:
Thanks for the updates, including the help and qldoc.  I think we still need:
- [ ] an answer to: "Shouldn't that be just Argument[-1].Element[@] then (i.e. 0 or more indirections - not 1 or more indirections)?".
- [ ] after that I need to review the new models.
- [ ] change note for the newly supported MAD syntax.

Then I think we can merge.

MathiasVP commented:
Yep, (1) and (3) are coming later today 👍 Thanks for summarizing it!

geoffw0 commented on cpp/ql/lib/ext/std.forward_list.model.yml:
```diff
@@ -0,0 +1,55 @@
+      - ["std", "forward_list<T>", True, "insert_after", "(const_iterator,size_type,const T &)", "", "Argument[*@2]", "Argument[-1].Element[@]", "value", "manual"]
+      - ["std", "forward_list<T>", True, "insert_after", "(const_iterator,size_type,const T &)", "", "Argument[*@2]", "ReturnValue.Element[@]", "value", "manual"]
+      - ["std", "forward_list<T>", True, "insert_after", "(const_iterator,T &&)", "", "Argument[*@1]", "Argument[-1].Element[@]", "value", "manual"]
+      - ["std", "forward_list<T>", True, "insert_after", "(const_iterator,T &&)", "", "Argument[*@1]", "ReturnValue.Element[@]", "value", "manual"]
```
Should we have a version of this model that flows from `Argument[-1].Element[@]` to `ReturnValue.Element[@]` (as we do for the other `insert_after` models)?

The same applies to `deque::insert`, `list::insert` and `vector::insert`.

MathiasVP replied:
Yeah, we definitely should. I've added this in f9d6c63cbb353c6f921faa12b086db09a973a0ec

geoffw0 commented on cpp/ql/lib/ext/std.vector.model.yml:
```diff
@@ -0,0 +1,59 @@
+      extensible: summaryModel
+    data: # namespace, type, subtypes, name, signature, ext, input, output, kind, provenance
+      - ["std", "vector", True, "assign", "", "", "Argument[*@1]", "Argument[-1].Element[@]", "value", "manual"]
+      - ["std", "vector", True, "assign<InputIt>", "(InputIt,InputIt)", "", "Argument[0]", "Argument[-1].Element[@]", "value", "manual"]
```
Should this be from `Argument[0].Element[@]`, since in this case the argument is an iterator?  (same applies for `list`, `forward_list` and `deque`).

MathiasVP replied:
Indeed, they should! Fixed in 5351c2734f551870fbbe90222af06ea2e71f3787

geoffw0 commented on cpp/ql/lib/ext/std.vector.model.yml:
```diff
@@ -0,0 +1,59 @@
+      - ["std", "vector", True, "cbegin", "", "", "Argument[-1].Element[@]", "ReturnValue.Element[@]", "value", "manual"]
+      - ["std", "vector", True, "data", "", "", "Argument[-1].Element[@]", "ReturnValue[*@]", "value", "manual"]
+      - ["std", "vector", True, "emplace", "", "", "Argument[*@1]", "Argument[-1].Element[@]", "value", "manual"]
+      - ["std", "vector", True, "emplace", "", "", "Argument[*@1]", "ReturnValue[*@]", "value", "manual"]
```
Should the `emplace` models flow to `ReturnValue.Element[@]`, since the return value is an iterator?

MathiasVP replied:
Well spotted! Fixed in 6d053247244c9bc3e4db727451a0edc96a076001. There are some fun discrepancies where some `emplace`-like functions return [references](https://en.cppreference.com/w/cpp/container/vector/emplace_back), and others return [iterators](https://en.cppreference.com/w/cpp/container/vector/emplace). I _think_ I've gotten them right now 🤞

geoffw0 commented on cpp/ql/lib/ext/std.deque.model .yml:
```diff
@@ -0,0 +1,72 @@
+extensions:
```
This file is named `cpp/ql/lib/ext/std.deque.model .yml` (spurious space).

MathiasVP replied:
Fixed in e03f8084e62f087bfed247ad7c9e39b4202fb51d

MathiasVP commented on cpp/ql/lib/semmle/code/cpp/models/implementations/StdContainer.qll:
```diff
@@ -55,73 +55,6 @@ private class Vector extends StdSequenceContainer {
-    // reverse flow from returned reference to the qualifier (for writes to
-    // `data`)
-    input.isReturnValueDeref() and
-    output.isQualifierObject()
```
Correct :) Inside the dataflow library, this is the exact same mechanism as the one that gives us flow out of `set` and into `s.x`'s post update node in:
```cpp
void set(int* p) { *p = 42; }

struct S { int x; };
void test() {
  S s;
  set(&s.x);
}
```

MathiasVP commented on cpp/ql/lib/ext/std.array.model.yml:
```diff
@@ -0,0 +1,12 @@
+      - ["std", "array", True, "data", "", "", "Argument[-1].Element[@]", "ReturnValue[*@]", "value", "manual"]
+      - ["std", "array", True, "operator[]", "", "", "Argument[-1].Element[@]", "ReturnValue[*@]", "value", "manual"]
+      - ["std", "array", True, "rbegin", "", "", "Argument[-1].Element[@]", "ReturnValue.Element[@]", "value", "manual"]
+      - ["std", "array", True, "rcbegin", "", "", "Argument[-1].Element[@]", "ReturnValue.Element[@]", "value", "manual"]
```
Fixed in c4dabb94d6ea8626382fc71f6d851ad08740f724

geoffw0 replied:
Still missing for `bsl`.

MathiasVP replied:
And _that_ is now fixed in af28dd8eb4c85e4a9ab14280f3e2540dcd57fb58. Thanks!

MathiasVP commented:
I _think_ I've addressed all your 🦅 eyed comments on the models now 🤞

geoffw0 reviewed (approved):
Everything LGTM now. :tada:
